lib/Dialect/TTNN/IR/TTNNOps.cpp

```diff
@@ -107,6 +107,57 @@ ::mlir::LogicalResult mlir::tt::ttnn::TransposeOp::verify() {
   return success();
 }
 
+::mlir::LogicalResult mlir::tt::ttnn::ReshapeOp::verify() {
+  ::mlir::RankedTensorType inputType = getInput().getType();
+  ::mlir::RankedTensorType outputType = getOutput().getType();
+  ArrayAttr shape = getShape();
+  int64_t shape_size = static_cast<int64_t>(shape.size());
+
+  if (shape_size == 0) {
+    return emitOpError("Shape attribute must be non-empty");
+  }
+
+  if (shape_size > 5) {
+    return emitOpError("Shape attribute must have at most 5 elements");
+  }
+
+  if (inputType.getNumElements() != outputType.getNumElements()) {
+    return emitOpError(
+        "Input and output tensors must have the same number of elements");
+  }
+
+  bool has_negative = false;
+  for (int64_t i = 0; i < shape_size; i++) {
+    if (mlir::cast<IntegerAttr>(shape[i]).getInt() == -1) {
+      if (has_negative) {
+        return emitOpError("Shape attribute must have at most one -1 element");
+      }
+
+      has_negative = true;
+    }
+  }
+
```

Comment:
Add a check that compares volume (num elements) between shape_size and output tensor volume, something like this:
```
if (!has_negative && shape_size.product() != outputType.getNumElements())
{
  return emitOpError("Shape attribute volume must match output tensor volume.");
}
```

(`.product()` doesn't exist, you can calculate that beforehand)

Reply:
And then, if there is a `-1`, we should check for divisibility:
```
if (has_negative && (outputType.getNumElements % shape_size.product_without_minus_one() != 0))
{
  return emitOpError(...);
}